Update bundled GoogleTest version to 1.14.0 and use FetchContent to build it

dalg24 opened the pull request:
* Update bundled GoogleTest from unknown very old to v1.14.0
  Googletest v1.14.0 is distributed with Ubuntu 24.04 LTS (Ubuntu 22.04 comes with 1.11.0)
  For reference, we also picked a minimum version of Google Benchmark that is aligned with Ubuntu 24.04
* Specify a minimum required version
* Use FetchContent to build GoogleTest (and GoogleMock)

dalg24 commented:
I downloaded GoogleTest v1.14.0 from https://github.com/google/googletest/releases/tag/v1.14.0 and added the whole directory after expanding the archive.

dalg24 commented:
Apparently support for `.clang-format-ignore` files was only added in Clang-Format 18.1.0

romintomasetti commented:
:partying_face:  This makes me happy because you're bringing in `gmock` - that means a lot of test assertions can become much cleaner (thinking of strings for instance). For instance you will be able to update:
https://github.com/kokkos/kokkos/blob/aecc5dc8f5be7df3a4d8b9c6fa99f1212475bccc/core/unit_test/cuda/TestCuda_InterOp_Graph.cpp#L113-L116

to a more self-explanatory:
```c++
EXPECT_THAT(buffer.str(), ::testing::MatchesRegex(expected));
```

:slightly_frowning_face: That said, I’m a bit puzzled by the decision not to use a *submodule*. This PR adds or changes over 200 files in the repository, which feels quite heavy. Moreover, `Kokkos` **never patches** `gtest`, which makes it even more questionable to hard-copy it into the tree.

I know the answer is probably historical — that `Kokkos` avoids submodules to keep things simpler for new users — but in practice, `FetchContent` or a submodule both involve fetching external dependencies. If the user doesn’t clone `Kokkos` with the submodule, `FetchContent` will fetch it anyway. So the complexity seems similar, and a submodule would keep the repo much leaner. Curious what your thoughts are on that trade-off?

tretre91 commented on cmake/kokkos_tribits.cmake:
```diff
@@ -6,12 +6,23 @@ include(GNUInstallDirs)
+    message(STATUS "Using bundled GoogleTest copy (version \"1.14.0\")")
+
+    # Suppress clang-tidy diagnostics on code that we do not have control over
+    if(CMAKE_CXX_CLANG_TIDY)
+      set_target_properties(gtest gtest_main gmock gmock_main PROPERTIES CXX_CLANG_TIDY "")
+    endif()
```
You might want to also suppress warnings for these targets (we do this already for google benchmark)
```suggestion
    if(CMAKE_CXX_CLANG_TIDY)
      set_target_properties(gtest gtest_main gmock gmock_main PROPERTIES CXX_CLANG_TIDY "")
    endif()
    
    set_target_properties(gtest gtest_main gmock gmock_main PROPERTIES COMPILE_OPTIONS -w)
```

dalg24 replied:
Applied.  Thanks

tretre91 commented on cmake/kokkos_tribits.cmake:
```diff
@@ -6,12 +6,23 @@ include(GNUInstallDirs)
+    if(CMAKE_CXX_CLANG_TIDY)
+      set_target_properties(gtest gtest_main gmock gmock_main PROPERTIES CXX_CLANG_TIDY "")
+    endif()
+  endif()
```
This should be put somewhere after the call to `kokkos_setup_build_environment` in the root CMakeLists.txt, otherwise it fails when `Kokkos_INSTALL_TESTING` is enabled because we check for `Kokkos_ENABLE_TESTS` before it has actually been set.
Or maybe do `if (Kokkos_ENABLE_TESTS OR Kokkos_INSTALL_TESTING)`.

tpadioleau commented:
> 🙁 That said, I’m a bit puzzled by the decision not to use a submodule. This PR adds or changes over 200 files in the repository, which feels quite heavy.

I kind of agree.

> Moreover, Kokkos never patches gtest, which makes it even more questionable to hard-copy it into the tree.

I think we actually did, looking at the history of `gtest-all.cc` we have some bug fixes regarding C++20 for example. That makes me wonder if we should also have the same CI check as for mdspan and desul to be sure we stick to the source code of the version we claim to embed ?

Also I think we need to set `INSTALL_GTEST=OFF` to disable the installation of GoogleTest during the installation of Kokkos.

dalg24 commented:
> > Moreover, Kokkos never patches gtest, which makes it even more questionable to hard-copy it into the tree.
> 
> I think we actually did, looking at the history of `gtest-all.cc` we have some bug fixes regarding C++20 for example. That makes me wonder if we should also have the same CI check as for mdspan and desul to be sure we stick to the source code of the version we claim to embed ?

We did plenty of patching :/
I am ok with adding a CI check.  Would you want it as part of this PR (assuming that we end up agreeing on bundling a copy of gtest)?

dalg24 commented:
> > 🙁 That said, I’m a bit puzzled by the decision not to use a submodule. This PR adds or changes over 200 files in the repository, which feels quite heavy.
> 
> I kind of agree.

Discussions around use of submodules or needing to retrieve sources from the internet have been controversial in the past.
What I really want to achieve in this PR is two-fold: (1 and foremost) cleanup our build system by making googletest treatment closer to what we would do with a regular dependency, and (2) enable using googlemock and more modern features in gtest.

I am proposing to bundle a copy because I expect that has more chances to get consensus.

masterleinad commented:
> I am proposing to bundle a copy because I expect that has more chances to get consensus.

I'd prefer to use FetchContent_Declare like we do for benchmarks (and switch to a different strategy when/if we actually need to apply patches).

dalg24 commented:
OK I heard the feedback about preferring to download an archive rather than bundling the copy.
Let us table that discussion for now and get everything else to pass.
We can revive it when it is ready.

dalg24 commented:
I changed it to download the archive from GH for now.
I checked that `-DFETCHCONTENT_SOURCE_DIR_GOOGLETEST=/path/to/gtest/source/dir` bypass the download.
Obviously `-DGTest_ROOT=/path/to/gtest/install/dir` is also an option.

masterleinad commented:
Looks like we need the patch in https://github.com/google/googletest/issues/3554#issuecomment-2636883816 for making Windows with shared libraries work.

romintomasetti commented on cmake/kokkos_tribits.cmake:
```diff
@@ -6,12 +6,30 @@ include(GNUInstallDirs)
-  set(KOKKOS_GTEST_LIB kokkos_gtest)
-  message(STATUS "Using internal gtest for testing")
+if(Kokkos_ENABLE_TESTS OR Kokkos_INSTALL_TESTING)
+  find_package(GTest QUIET 1.14.0)
```
Please motivate your choice for the version, i.e. why not choosing 1.15 or 1.16 or 1.17?

romintomasetti replied:
(In the PR description plz)

dalg24 replied:
Done

dalg24 committed:
Remove old tpls/gtest directory

Signed-off-by: Damien L-G <dalg24@gmail.com>

dalg24 committed:
Add googletest v1.14.0

Signed-off-by: Damien L-G <dalg24@gmail.com>

dalg24 committed:
Make FetchContent point to newer GoogleTest

Signed-off-by: Damien L-G <dalg24@gmail.com>

dalg24 committed:
Let GoogleTest v1.14.0 be the minimum required

Signed-off-by: Damien L-G <dalg24@gmail.com>

dalg24 committed:
fixup! Add googletest v1.14.0

dalg24 committed:
fixup! Add googletest v1.14.0

dalg24 committed:
Disable installing GTest

Signed-off-by: Damien L-G <dalg24@gmail.com>

dalg24 committed:
Merge branch 'develop' of https://github.com/kokkos/kokkos into gtest_tpl

dalg24 committed:
Remove files configured in install

Signed-off-by: Damien L-G <dalg24@gmail.com>

dalg24 committed:
Disable cmake-format in tpls/

Signed-off-by: Damien L-G <dalg24@gmail.com>

dalg24 committed:
CMake: Do not enable the C language (unnecessary)

The project() command in the top-level CMakeLists.txt was not specifying
languages which default to enabling C and CXX.
The googlemock and googletest project() commands were explicitly enabing
both the C and CXX languages in their respective subdirectories,
although all their source file are correctly identified as CXX and the C
language is effectively never used.

Enabling the C language forces CXX-only projects downstream that embed
this repo (via FetchContent or add subdirectory) to detect a C compiler
at configuration-time.

This PR proposes to properly only enable the CXX language.

Signed-off-by: Damien L-G <dalg24@gmail.com>

dalg24 committed:
Try to fix Windows builds

Signed-off-by: Damien L-G <dalg24@gmail.com>

dalg24 committed:
Pre-install GoogleTest in build with Kokkos_INSTALL_TESTING

Signed-off-by: Damien L-G <dalg24@gmail.com>

dalg24 committed:
Merge branch 'develop' of https://github.com/kokkos/kokkos into gtest_tpl

dalg24 committed:
Fixup DISABLE -> REQUIRE

Signed-off-by: Damien L-G <dalg24@gmail.com>

cwpearson commented:
I would prefer this bundling approach if there is any convenience benefit to folks who work with Kokkos in air-gapped environments. I believe the maintenance burden of either is essentially the same for us, so we should choose the option that helps the users the most.

cedricchevalier19 commented:
Can you explain a little bit more the need for patches?
If Kokkos use an installed GTest, found by `find_package` do we need patches? If not, why not using `ExternalProject` instead of `FetchContent` to have a more separated build?

Rombur committed:
Ignore undef warnings

Signed-off-by: Bruno Turcksin <bruno.turcksin@gmail.com>

Rombur committed:
Use bundled gtest and ignore undef warnings in Jenkins

Signed-off-by: Bruno Turcksin <bruno.turcksin@gmail.com>

Rombur committed:
Ignore warnings about extra parenthesess

Signed-off-by: Bruno Turcksin <bruno.turcksin@gmail.com>

Rombur committed:
Fix typo

Signed-off-by: Bruno Turcksin <bruno.turcksin@gmail.com>

Rombur committed:
Ignore warnings about extra parenthesess in jenkins

Signed-off-by: Bruno Turcksin <bruno.turcksin@gmail.com>

Rombur committed:
Disable more warnings

Signed-off-by: Bruno Turcksin <bruno.turcksin@gmail.com>

Rombur committed:
Trying to fix sycl and nvhpc

Signed-off-by: Bruno Turcksin <bruno.turcksin@gmail.com>

Rombur committed:
Trying to fix sycl

Signed-off-by: Bruno Turcksin <bruno.turcksin@gmail.com>

Rombur committed:
Trying to fix nvhpc

Signed-off-by: Bruno Turcksin <bruno.turcksin@gmail.com>

Rombur committed:
Merge branch 'develop' into gtest_tpl

Rombur committed:
Trying to fix nvhpc

Signed-off-by: Bruno Turcksin <bruno.turcksin@gmail.com>

crtrott commented:
> Can you explain a little bit more the need for patches? If Kokkos use an installed GTest, found by `find_package` do we need patches? If not, why not using `ExternalProject` instead of `FetchContent` to have a more separated build?

Generally, Kokkos works with SOME compilers on SOME toolchains with an external unmodified GTEST, we absolutely do need patches to make it work with all our supported platforms.

crtrott commented on .jenkins/continuous.groovy:
```diff
@@ -238,7 +238,7 @@ pipeline {
                                 -DCMAKE_CXX_COMPILER_LAUNCHER=ccache \
                                 -DCMAKE_CXX_COMPILER=$WORKSPACE/bin/nvcc_wrapper \
-                                -DCMAKE_CXX_FLAGS="-Werror --Werror=all-warnings -Xcudafe --diag_suppress=940" \
+                                -DCMAKE_CXX_FLAGS="-Werror --Werror=all-warnings -Wno-error=undef -Wno-error=parentheses -Xcudafe --diag_suppress=940" \
```
Not happy with us to disable warnings globally for gtest.

crtrott committed:
GTEST-PATCH: Attempt to fix windows build

Signed-off-by: Christian Trott <crtrott@sandia.gov>

crtrott commented:
Ok applying the changes from the issue @masterleinad linked to, fixed the Windows link issues. Now its failing the tests with an exception thrown ...

Rombur committed:
Trying to fix nvhpc warning

Signed-off-by: Bruno Turcksin <bruno.turcksin@gmail.com>

Rombur committed:
Trying to fix nvhpc warning

Signed-off-by: Bruno Turcksin <bruno.turcksin@gmail.com>

Rombur committed:
Trying to fix nvhpc warning

Signed-off-by: Bruno Turcksin <bruno.turcksin@gmail.com>

Rombur committed:
Trying to fix nvhpc warning

Signed-off-by: Bruno Turcksin <bruno.turcksin@gmail.com>

Rombur commented:
I am stuck  trying to silence warnings of the type
```
[CUDA-12.2-NVHPC-AS-HOST-COMPILER] /var/jenkins/workspace/Kokkos_PR-8593/core/unit_test/cuda/TestCuda_InterOp_StreamsMultiGPU.cpp(57): error #177-D: member "<unnamed>::cuda_multi_gpu_managed_views_Test::test_info_" was declared but never referenced
```
I've tried all the flag combinations  I could think of but it didn't work. I think the warning comes from line 1544 in `tpls/googletest-1.14.0/googletest/include/gtest/internal/gtest-internal.h`
```
 static ::testing::TestInfo* const test_info_ GTEST_ATTRIBUTE_UNUSED_;  
```
I tried forcing `__attribute__((unused))` and using `[[maybe_unused]]`  but it doesn't work

masterleinad commented:
It's seems to only thing we can do here (without disabling warnings as errors) is prepending
```
#ifdef __NVCOMPILER
#pragma nv_diag_suppress 177
#endif
```
before any `gtest/gtest.h` includes (and we can't pop that suppression immediately after the include but only at the end of the file). We could just wrap the `gtest` header with that suppression and use that wrapper file everywhere instead (assuming we don't care about ever pushing/popping).

crtrott commented:
So far no luck for me reproducing locally on my windows box.

masterleinad commented:
GTest 1.18.0 has the same issue with NVHPC.

dalg24 commented on .github/workflows/continuous-integration-windows.yml:
```diff
@@ -37,10 +37,14 @@ jobs:
 
     steps:
     - uses: actions/checkout@3d3c42e5aac5ba805825da76410c181273ba90b1 # v7.0.1
+    - name: install gtest:x64-windows
```
```suggestion
    - name: install gtest
```

dalg24 commented on tpls/googletest-1.14.0/googletest/include/gtest/gtest.h:
```diff
@@ -0,0 +1,2331 @@
+#ifdef __NVCOMPILER
+#pragma nv_diagnostic push
+#pragma nv_diag_suppress 177
+#endif
```
Note this

dalg24 commented on tpls/googletest-1.14.0/googletest/include/gtest/gtest.h:
```diff
@@ -0,0 +1,2331 @@
+#ifdef __NVCOMPILER
+#pragma nv_diagnostic pop
+#pragma nv_diag_suppress 177
+#endif
```
And here

dalg24 commented on tpls/googletest-1.14.0/googletest/include/gtest/gtest.h:
```diff
@@ -0,0 +1,2331 @@
+
+#ifdef __NVCOMPILER
+#pragma nv_diagnostic pop
+#pragma nv_diag_suppress 177
```
```suggestion
```

dalg24 commented on .gitlab/hpsf-gitlab-ci.yml:
```diff
@@ -29,7 +29,7 @@ NVIDIA-RTX5080:
     - export ENV_CMAKE_OPTIONS="${ENV_CMAKE_OPTIONS};-D CMAKE_BUILD_TYPE=${BUILD_TYPE}"
     - export ENV_CMAKE_OPTIONS="${ENV_CMAKE_OPTIONS};-D CMAKE_CXX_COMPILER=`pwd`/bin/nvcc_wrapper"
-    - export ENV_CMAKE_OPTIONS="${ENV_CMAKE_OPTIONS};-D CMAKE_CXX_FLAGS='-Werror=all-warnings -Werror'"
+    - export ENV_CMAKE_OPTIONS="${ENV_CMAKE_OPTIONS};-D CMAKE_CXX_FLAGS='-Werror=all-warnings -Werror -Wno-error=undef -Wno-error=parentheses'"
```
Why do we need these?  Are these for NVCC?

dalg24 replied:
Can we achieve it with pragma push/pop?

masterleinad commented on cmake/kokkos_tribits.cmake:
```suggestion
    set_target_properties(gtest PROPERTIES COMPILE_OPTIONS -w)
    set_target_properties(gtest_main PROPERTIES COMPILE_OPTIONS -w)
    set_target_properties(gmock PROPERTIES COMPILE_OPTIONS -w)
    set_target_properties(gmock_main PROPERTIES COMPILE_OPTIONS -w)
```

crtrott commented on .gitlab/hpsf-gitlab-ci.yml:
```diff
@@ -29,7 +29,7 @@ NVIDIA-RTX5080:
     - export ENV_CMAKE_OPTIONS="${ENV_CMAKE_OPTIONS};-D CMAKE_BUILD_TYPE=${BUILD_TYPE}"
     - export ENV_CMAKE_OPTIONS="${ENV_CMAKE_OPTIONS};-D CMAKE_CXX_COMPILER=`pwd`/bin/nvcc_wrapper"
-    - export ENV_CMAKE_OPTIONS="${ENV_CMAKE_OPTIONS};-D CMAKE_CXX_FLAGS='-Werror=all-warnings -Werror'"
+    - export ENV_CMAKE_OPTIONS="${ENV_CMAKE_OPTIONS};-D CMAKE_CXX_FLAGS='-Werror=all-warnings -Werror -Wno-error=undef -Wno-error=parentheses'"
```
Why do we need this here but not for the SNL ones?

dalg24 replied:
Dunno.  @masterleinad did you add these?

masterleinad replied:
I readded those in the Jenkins file again in https://github.com/kokkos/kokkos/pull/8593/changes/312694ed850eae21e7f5885bd47f780abd8102ac after errors in https://cloud1.cees.ornl.gov/jenkins-ci/job/Kokkos/job/PR-8593/37/stages/.

masterleinad commented on .github/workflows/continuous-integration-windows.yml:
```diff
@@ -37,10 +37,14 @@ jobs:
       shell: bash
       run: |
-        cmake -B build -DBUILD_SHARED_LIBS=ON -DKokkos_ENABLE_SERIAL=ON -DKokkos_ENABLE_TESTS=ON
+        cmake -B build -DBUILD_SHARED_LIBS=ON -DKokkos_ENABLE_SERIAL=ON -DKokkos_ENABLE_TESTS=ON -DCMAKE_TOOLCHAIN_FILE=$VCPKG_INSTALLATION_ROOT/scripts/buildsystems/vcpkg.cmake -DCMAKE_REQUIRE_FIND_PACKAGE_GTest=ON
```
Why do we need `CMAKE_TOOLCHAIN_FILE` here. What happens if we omit it?

dalg24 replied:
We need to tell CMake where to find GTest and pointing to the vcpkg toolchain file is what is being recommended https://learn.microsoft.com/en-us/vcpkg/users/buildsystems/cmake-integration

masterleinad replied:
AI is telling me:

The vcpkg.cmake toolchain file acts as a bridge that automatically injects dependency resolution logic into the CMake build environment. It intercepts built-in CMake functions during the initialization phase to ensure your project seamlessly locates libraries managed by vcpkg. 
The file alters and controls the CMake build environment through several key mechanisms:[1](https://learn.microsoft.com/en-us/vcpkg/users/buildsystems/cmake-integration). 

1. Automatic Dependency Resolution (Manifest Mode)
If your project directory contains a vcpkg.json manifest file, the toolchain will automatically trigger vcpkg install before your project configures. This downloads, builds, and prepares all required dependencies behind the scenes without manual user intervention.[2](https://www.youtube.com/watch?v=10kqdZR4rmU&vl=en&t=119). 

2. Modification of Search Paths
The toolchain adjusts CMake search paths to prioritize the folders containing your installed vcpkg packages. It appends or prepends vcpkg target directories to the following standard CMake variables:CMAKE_PREFIX_PATH (for locating package config files)CMAKE_LIBRARY_PATH (for locating binaries and .lib/.a files)CMAKE_FIND_ROOT_PATH (for target-redirected filesystem lookups)This modification ensures that standard CMake commands like find_package(), find_library(), and find_path() will immediately discover your vcpkg packages.[3](https://www.reddit.com/r/cpp_questions/comments/1769mwf/how_to_properly_use_vcpkg_for_libraries_that_seem/). 

3. Target Triplet Auto-Detection
The toolchain file evaluates your project's active compiler settings to automatically select an appropriate target triplet (e.g., x64-windows, x64-linux). This informs CMake exactly which architecture, platform, and linking configuration (static vs. dynamic) it should consume from the vcpkg installation.
  
4. Overriding Target Commands
To streamline deployment, the toolchain intercepts commands like add_executable(). On platforms like Windows, it appends a hidden POST_BUILD step to targets. This step automatically copies required shared libraries (.dll files) into your output build directory alongside your executable so it can run immediately.[5](https://learn.microsoft.com/en-us/vcpkg/users/buildsystems/cmake-integration). 
 
5. Managing Custom Toolchains (Chainloading)
Because CMake only permits a single CMAKE_TOOLCHAIN_FILE at a time, the vcpkg toolchain includes a mechanism called chainloading. If you need a custom toolchain to define your compilers or flags, you pass it to vcpkg using the VCPKG_CHAINLOAD_TOOLCHAIN_FILE variable. The vcpkg toolchain will execute your custom file first before overlaying its own package management environment.

masterleinad replied:
That means we should be able to avoid https://github.com/kokkos/kokkos/blob/2b66fce67a4ba10d697aaa9be5f12abcd2f910d4/.github/workflows/continuous-integration-windows.yml#L51 if we were to use it for the Cuda + shared libraries build.

masterleinad commented on tpls/googletest-1.14.0/googletest/include/gtest/gtest.h:
```diff
@@ -49,11 +49,6 @@
-#ifdef __NVCOMPILER
-#pragma nv_diagnostic push
-#pragma nv_diag_suppress 177
-#endif
```
We've been seeing these issues when compiling Kokkos tests not when compiliing `gtest` so I'm skeptical if we can just drop this.

dalg24 commented on cmake/kokkos_tribits.cmake:
```diff
@@ -7,14 +7,19 @@ include(GNUInstallDirs)
     list(APPEND CMAKE_MESSAGE_INDENT "[googletest] ")
-    FetchContent_Declare(googletest SOURCE_DIR ${Kokkos_SOURCE_DIR}/tpls/gtest)
+    set(INSTALL_GTEST OFF)
+    # Kokkos only ever links against GTest::gtest
+    set(BUILD_GMOCK OFF)
```
Intended to be re-enabled when we actually start using it but could in principle be done now with the set_target_properties trick below.

masterleinad reviewed (changes requested):
I'm not convinced that the current status addresses the Cuda + nvc++ (host compiler) issues we have been struggling with for the longest time in this pull request.

tcclevenger reviewed (approved):
LGTM

dalg24 commented:
> I'm not convinced that the current status addresses the Cuda + nvc++ (host compiler) issues we have been struggling with for the longest time in this pull request.

Since the ORNL Jenkins CI does not have test machines with NVIDIA GPUs online at the moment, please test in a local docker build and report back.

masterleinad commented:
> Since the ORNL Jenkins CI does not have test machines with NVIDIA GPUs online at the moment, please test in a local docker build and report back.

Yes, it's still failing.
> /app/kokkos/core/unit_test/cuda/TestCuda_InterOp_StreamsMultiGPU.cpp(108): error #177-D: member "<unnamed>::cuda_multi_gpu_stream_sync_semantics_raw_cuda_Test::test_info_" was declared but never referenced
  static_assert(sizeof("cuda_multi_gpu") > 1, "test_suite_name must not be empty"); static_assert(sizeof("stream_sync_semantics_raw_cuda") > 1, "test_name must not be empty"); class cuda_multi_gpu_stream_sync_semantics_raw_cuda_Test : public cuda_multi_gpu { public: cuda_multi_gpu_stream_sync_semantics_raw_cuda_Test() = default; ~cuda_multi_gpu_stream_sync_semantics_raw_cuda_Test() override = default; cuda_multi_gpu_stream_sync_semantics_raw_cuda_Test (const cuda_multi_gpu_stream_sync_semantics_raw_cuda_Test &) = delete; cuda_multi_gpu_stream_sync_semantics_raw_cuda_Test & operator=( const cuda_multi_gpu_stream_sync_semantics_raw_cuda_Test &) = delete; cuda_multi_gpu_stream_sync_semantics_raw_cuda_Test (cuda_multi_gpu_stream_sync_semantics_raw_cuda_Test &&) noexcept = delete; cuda_multi_gpu_stream_sync_semantics_raw_cuda_Test & operator=( cuda_multi_gpu_stream_sync_semantics_raw_cuda_Test &&) noexcept = delete; private: void TestBody() override; static ::testing::TestInfo* const test_info_ __attribute__((unused)); }; ::testing::TestInfo* const cuda_multi_gpu_stream_sync_semantics_raw_cuda_Test::test_info_ = ::testing::internal::MakeAndRegisterTestInfo( "cuda_multi_gpu", "stream_sync_semantics_raw_cuda", nullptr, nullptr, ::testing::internal::CodeLocation("/app/kokkos/core/unit_test/cuda/TestCuda_InterOp_StreamsMultiGPU.cpp", 108), (::testing::internal::GetTypeId<cuda_multi_gpu>()), ::testing::internal::SuiteApiResolver< cuda_multi_gpu> ::GetSetUpCaseOrSuite("/app/kokkos/core/unit_test/cuda/TestCuda_InterOp_StreamsMultiGPU.cpp", 108), ::testing::internal::SuiteApiResolver< cuda_multi_gpu> ::GetTearDownCaseOrSuite("/app/kokkos/core/unit_test/cuda/TestCuda_InterOp_StreamsMultiGPU.cpp", 108), new ::testing::internal::TestFactoryImpl<cuda_multi_gpu_stream_sync_semantics_raw_cuda_Test>); void cuda_multi_gpu_stream_sync_semantics_raw_cuda_Test::TestBody() {

We need
```diff
diff --git a/tpls/googletest-1.14.0/googletest/include/gtest/gtest.h b/tpls/googletest-1.14.0/googletest/include/gtest/gtest.h
index de7d528fc..b2dfc2dd3 100644
--- a/tpls/googletest-1.14.0/googletest/include/gtest/gtest.h
+++ b/tpls/googletest-1.14.0/googletest/include/gtest/gtest.h
@@ -49,6 +49,10 @@
 #ifndef GOOGLETEST_INCLUDE_GTEST_GTEST_H_
 #define GOOGLETEST_INCLUDE_GTEST_GTEST_H_
 
+#ifdef __NVCOMPILER
+#pragma nv_diag_suppress 177
+#endif
+
 #include <cstddef>
 #include <cstdint>
 #include <iomanip>
```
to avoid the warnings. In particular, we can't push/pop since the suppression needs to stay active throughout the declaration of the test cases.
